app/code/Magento/Config/Model/Config/Source/Locale/Currency.php

```diff
@@ -21,27 +24,70 @@ class Currency implements \Magento\Framework\Option\ArrayInterface
     protected $_options;
 
     /**
-     * @var \Magento\Framework\Locale\ListsInterface
+     * @var ListsInterface
      */
     protected $_localeLists;
 
     /**
-     * @param \Magento\Framework\Locale\ListsInterface $localeLists
+     * @var ScopeConfigInterface
      */
-    public function __construct(\Magento\Framework\Locale\ListsInterface $localeLists)
-    {
+    private $config;
+
+    /**
+     * @var array
+     */
+    private $installedCurrencies;
+
+    /**
+     * @param ListsInterface $localeLists
+     * @param ScopeConfigInterface $config
+     */
+    public function __construct(
+        ListsInterface $localeLists,
+        ScopeConfigInterface $config = null
+    ) {
         $this->_localeLists = $localeLists;
+        $this->config = $config ?: ObjectManager::getInstance()->get(ScopeConfigInterface::class);
     }
 
     /**
-     * @return array
+     * @inheritdoc
      */
     public function toOptionArray()
     {
         if (!$this->_options) {
             $this->_options = $this->_localeLists->getOptionCurrencies();
         }
-        $options = $this->_options;
+
+        $selected = $this->getInstalledCurrencies();
+
+        $options = array_filter(
+            $this->_options,
+            function ($option) use ($selected) {
+                return in_array($option['value'], $selected);
```

Comment:
Please array so that we can check with `isset`.

Reply:
> Please array so that we can check with `isset`.

Hi orlangur!
Please take a look if this is what you mean:
https://github.com/magento/magento2/pull/22230/commits/de3d22e77f15d85437a35c112bc1cadfa3a52ca8

Reply:
@melaxon,

>  If one day some exotic currency will be used this checking may fail

I don't see how it can fail, also, 3 letters is an ISO standard.

No need in filling keys, just use `array_flip`.

Reply:
> ..., just use `array_flip`.

Done: https://github.com/magento/magento2/pull/22230/commits/2d9465ceb996de0f6b8a9a9b98905470e9f05517